### ISSUE TITLE
[QUEST] Glimmer

### ISSUE THREAD

**Issue by jgwhite** [jgwhite/prettier#1]
## Objective

We want to get Prettier’s support for Ember templates up to the point where it is complete and maintainable enough to be officially released.

This quest issue was created to track the outstanding work required to meet this objective and to help volunteers coordinate to get the work done.

## Contributing

The easiest way to contribute is to try Prettier’s Ember template support out in your apps and report any unwanted, incomplete, or otherwise wonky behaviour as a comment on this issue so it can be added to the quest. To do this:

```bash
$ yarn add -D prettier
$ yarn prettier '**/*.hbs' --parser glimmer
```

Or

```bash
$ npm i --save-dev prettier
$ npm run prettier '**/*.hbs' --parser glimmer
```

Then check the changes and look for stuff that makes you frown!

When you’re ready to contribute some code, identify an unowned item on the list below and leave a comment stating that you intend to work on it. We’ll tag the item with your handle. Once you have a PR that we can reference, leave another comment so we can link the item to the PR.

If you decide for whatever reason that you no longer wish to pursue the work, leave another comment and we’ll remove your handle from it. This is absolutely fine and you shouldn’t feel any regret if things don’t pan out.

## Tasks

Check [bugs on Prettier labeled with "lang:handlebars"](https://github.com/prettier/prettier/issues?q=is%3Aopen+label%3Atype%3Abug+label%3Alang%3Ahandlebars).

## Why does this quest issue live on a fork of prettier/prettier?

We’re trying it out as an experiment so we can easily add administrators on this issue.

**Comment by rwjblue** [comment]
RE: the entity escaping, we really need to avoid the escaping all together (see https://github.com/prettier/prettier/pull/4533#issuecomment-495666923 for more details)

**Comment by GavinJoyce** [comment]
I'm working on fixing the HTML entities here: https://github.com/prettier/prettier/pull/6234

**Comment by tchak** [comment]
@jgwhite if you add me to this fork it will be easier for me to copy the source markdown to the upstream issue

**Comment by GavinJoyce** [comment]
Perhaps it would be better to just have a single quest issue? Perhaps we should just use this one as we can have multiple editors? We could link to this issue from https://github.com/prettier/prettier/issues/4908

@jgwhite perhaps you could give me edit access to this issue too?

**Comment by tchak** [comment]
I just linked the main issue to this one

**Comment by dcyriller** [comment]
I've opened a PR to fix a bug: prettier is a bit eager regarding whitespaces between text and helpers https://github.com/prettier/prettier/pull/6239

```handlebars
// Input
{{#component propA}}
    for {{propB}}  do {{propC}} f
{{/component}}

// Output (Prettier stable)
{{#component propA}}
  for{{propB}}do{{propC}}f
{{/component}}

// Output (Prettier the PR)
{{#component propA}}
  for {{propB}}  do {{propC}} f
{{/component}}
```

@GavinJoyce I hope it won't conflict with what your current work 😅

**Comment by Alonski** [comment]
@dcyriller Do you rely on this whitespace? I personally would like a more verbose way to specify significant whitespace. Maybe I'm wrong 😅

**Comment by jgwhite** [comment]
@dcyriller added to the tracking info description. Nice one!

Personally I’d like Prettier to behave like this:

**Before:**

```hbs
<p>Hello   {{name}}   happy   {{age}}    birthday!</p>

<pre>
  Name: {{name}}
  Age:  {{age}}
</pre>
```

**After:**

```hbs
<p>Hello {{name}} happy {{age}} birthday!</p>

<pre>
  Name: {{name}}
  Age:  {{age}}
</pre>
```

i.e. don’t mess with whitespace instead `<pre>` tags.

**Edit:**

Though it now occurs to me that there’d be no way for Prettier to know about components like:

```hbs
<CodeBlock @format="plain">
  Name: {{name}}
  Age:  {{age}}
</CodeBlock>
```

**Comment by GavinJoyce** [comment]
> Though it now occurs to me that there’d be no way for Prettier to know about components like:

I think we'll need to add support for `{{!-- prettier-ignore --}}`:

```hbs
{{!-- prettier-ignore --}}
<CodeBlock @format="plain">
  Name: {{name}}
  Age:  {{age}}
</CodeBlock>
```

**Comment by dcyriller** [comment]
Reading your comments I understand we should have:
```handlebars
// Input
{{#component propA}}
    for {{propB}}  do {{propC}} f
{{/component}}

// Output (Prettier stable)
{{#component propA}}
  for{{propB}}do{{propC}}f
{{/component}}

// Output (Prettier the PR)
{{#component propA}}
  for {{propB}} do {{propC}} f
{{/component}}
```
I'll update my PR tomorrow to reflect that!

Also, we should implement `{{!-- prettier-ignore --}}` for handlebars (most probably in an other PR).

Thank you for your feedbacks!

EDIT: just saw that you opened a new PR two hours before mine @GavinJoyce 😬. So, I'll close it for now. Once yours is merged, I'll be able to rebase and open mine (if still necessary).

**Comment by GavinJoyce** [comment]
@dcyriller please feel free to land anything you like, I'm going to switch focus this weekend and try [an experiment to rebuild prettier's glimmer printer](https://github.com/GavinJoyce/prettier/pull/2) from first principles. I doubt that it will come to anything, but I think it will be useful in learning prettier's internals.

**Comment by jgwhite** [comment]
@GavinJoyce I sense a future emberconf talk in the making…

**Comment by ghost** [comment]
@GavinJoyce I'm in the peanut gallery here, but I'm curious if there is any overlap with @rwjblue 's [ember-template-recast](https://github.com/ember-template-lint/ember-template-recast#ember-template-recast). If anything another implementation to compare.

**Comment by jgwhite** [comment]
And also @code0100fun’s [ember-template-rewrite](https://github.com/q2ebanking/ember-template-rewrite).

**Comment by dcyriller** [comment]
I opened a PR to add trailing newline to hbs files: https://github.com/prettier/prettier/pull/6243 (my IDE adds a final newline on every opened files so I had git changes every time I opened a file 😛. Also, Prettier adds final newline for every other language.)

**Comment by jgwhite** [comment]
@dcyriller fantastic. This is one of the major roadblocks we ran into as well.

**Comment by jgwhite** [comment]
I wonder if it’d be worth announcing this quest in the Ember Times? It’d be awesome to get the community trying out the current glimmer support and shaking loose more todos for the quest.

**Comment by Alonski** [comment]
@jgwhite Good idea :) I will see if anyone can write something up for this.
Would you be willing to write a short section?

**Comment by laurmurclar** [comment]
Is there an expected timeline for this work? I'd love to work on "Improve curly/text spacing" but it'll probably be the weekend before I could make significant progress on it

**Comment by jgwhite** [comment]
> Is there an expected timeline for this work?

@laurmurclar no timeline, please jump in! 👍

> I'd love to work on "Improve curly/text spacing" but it'll probably be the weekend before I could make significant progress on it

I’ll tag that item with your name. @dcyriller are you happy with that?

**Comment by dcyriller** [comment]
> I’ll tag that item with your name. @dcyriller are you happy with that?

Sure thing. Also, I think there are a few things to fix around `curly/text spacing`.

**Comment by jgwhite** [comment]
Now this has appeared in the Ember Times I’ve added some more guidance to the readme. @GavinJoyce would you mind adding a line or two on how to use the Prettier playground for Glimmer?

**Comment by dcyriller** [comment]
FWIW I’ve run Prettier against two codebases. You can find these examples here: https://github.com/dcyriller/ember-template-lint-plugin-prettier#examples

**Comment by samselikoff** [comment]
Trying this out on some templates in EmberMap's codebase, very nice work y'all!

Biggest unwanted behavior is definitely (1) single blank newlines used for visual clarity and (2) single blank spaces between words and expressions.

Before:

```hbs
{{ui-video-hero
  video=model
  on-end=(action 'playNextVideo')
  on-play=(action 'turnOnAutoplay')
  on-pause=(action 'turnOffAutoplay')}}

{{#ui-container}}
  <main class="my-4 pb-2 lg:flex justify-between">
    <article class="w-full lg:w-3/5 flex-no-shrink">
      {{#ui-title style='headline'}}
        {{model.title}}
      {{/ui-title}}

      {{#ui-p data-test-id="series-description"}}
        {{model.description}}
      {{/ui-p}}

      {{video-extras clip=model belongsToSeries=true}}
    </article>

    {{#if screen.isLargeAndUp}}
      <div class="w-2/5 max-w-390">
        <div class="mb-4 pl-3">
          {{topics/topic/video/components/series-playlist
            series=model.series
            activeClip=model}}
        </div>
      </div>
    {{/if}}

  </main>
{{/ui-container}}

{{questions-pancake}}
```

After:

```hbs
{{ui-video-hero
  video=model
  on-end=(action "playNextVideo")
  on-play=(action "turnOnAutoplay")
  on-pause=(action "turnOffAutoplay")
}}
{{#ui-container}}
  <main class="my-4 pb-2 lg:flex justify-between">
    <article class="w-full lg:w-3/5 flex-no-shrink">
      {{#ui-title style="headline"}}
        {{model.title}}
      {{/ui-title}}
      {{#ui-p data-test-id="series-description"}}
        {{model.description}}
      {{/ui-p}}
      {{video-extras clip=model belongsToSeries=true}}
    </article>
    {{#if screen.isLargeAndUp}}
      <div class="w-2/5 max-w-390">
        <div class="mb-4 pl-3">
          {{topics/topic/video/components/series-playlist
            series=model.series
            activeClip=model
          }}
        </div>
      </div>
    {{/if}}
  </main>
{{/ui-container}}
{{questions-pancake}}
```

Everything else seems to be pretty spot on!

**Comment by laurmurclar** [comment]
Very sorry folks, haven't had much of a chance to work on this! I probably won't for a while either. Mind removing my name from that task? 😞

**Comment by jgwhite** [comment]
@laurmurclar no problem at all, and no need to apologize.

**Comment by chadian** [comment]
https://github.com/prettier/prettier/pull/6354

Addresses a bunch of the whitespace and line break issues. The changes are probably best reflected in the jest snapshots. I also added a test case `preserved-spaces-and-breaks.hbs` to try and capture a few different combinations and edge cases. Existing tests were changed to preserve intentional line breaks similar to what @samselikoff [mentions](https://github.com/jgwhite/prettier/issues/1#issuecomment-506986737):
> (1) single blank newlines used for visual clarity

[Give it a try](https://deploy-preview-6354--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEACVAeAJgSwG7qFgA2AhgM7kC8AOiAjAE4CedhqAfDVOhgBYBGDsGDoYOGMTjsAvjIwB6QVx6ZcBdCQrU6AIwhZWIFe1OERhfYbOE53Reo6YFjkABoQEAA7jo5ZKCkjIwQAO4ACkEI-iikeBA4WO4guoykYADWcDAAyl7pOFAA5shMAK5wHnwwALbEAOp8EnDk+WBwOdES+BLMyOCUyYXkcIww4WlFNaTIAGakxCMeAFbkAB4AQmmZ2TmkNXAAMoVwcwtLIKtrOYVFUgCKZRDwZ4uVIPmMI4z9dzg1Bx+Hi8jEKMHqiRgfGQAA4AAzAkIjeppLz9EEtUZ4U4eACOT3gE28MRAFAAtFA4HAsNTkow4PicPSJqQpjMkPM3h4RjUcKVGBVubcHgTThzzu8YKRdBCsFDkAAmDxMUg4Yi3ADCEAB7JAUGgOJAZRGABVpTFORc8BUAJJQGmwHJgUE+ACC9pyMGYUleIzkQA). I would love any feedback!

**Comment by samselikoff** [comment]
@chadian Very nice man! This takes care of all the issues I had last time I tried it 👏

Bug in my Atom editor integration that won't let me format-on-save, but once I do I think I'll make a video on this. Super exciting everyone, great stuff.

**Comment by FabHof** [comment]
I can have a look at the line breaks of dynamic element class names.

**Comment by dcyriller** [comment]
@jgwhite here are a few recent developments that you may want to reference in the description:

a small bug fix: https://github.com/prettier/prettier/pull/6377 (it applies `--single-quote` option in AttrNode)

a test only PR: https://github.com/prettier/prettier/pull/6378

a small issue to tackle (probably related to the unwanted line breaks within dynamic element class name): https://github.com/prettier/prettier/issues/6252

**Comment by jgwhite** [comment]
@dcyriller awesome. I’ve given you admin right on this fork — would you mind updating the issue description?

**Comment by samselikoff** [comment]
Quick question, has anybody been able to use this with format-on-save in Prettier?

**Comment by Alonski** [comment]
@samselikoff You use Atom right? There is a VSCode extension: https://marketplace.visualstudio.com/items?itemName=mfeckies.handlebars-formatter

**Comment by samselikoff** [comment]
Ya Atom

**Comment by FabHof** [comment]
I did some work on the spaces in dynamic element class name: prettier/prettier#6464

**Comment by chadian** [comment]
`Improve whitespace handling (@chadian)` has been merged. @jgwhite Can you check it off on the list for me? Thanks!

**Comment by xomaczar** [comment]
@jgwhite I think @chadian also fixed `Improve curly/text spacing` with his latest PR.

**Comment by samselikoff** [comment]
I'd like to try out the latest round of improvements! I'm using VSCode now but the two marketplace extensions have prettier pinned I believe, so I'm trying to get it setup on my own to no avail.

Basically tried installing `master` of Prettier and adding this to my project

```js
module.exports = {
  overrides: [
    {
      files: "*.hbs",
      options: { parser: "glimmer" }
    }
  ]
};
```

but code is telling me there's no formatter installed for 'handlebars' files. Anyone have any tips?

**Comment by jgwhite** [comment]
@samselikoff until Glimmer is officially supported you’ll need a prettier handlebars VSCode extension [like this one](https://marketplace.visualstudio.com/items?itemName=Alonski.prettier-for-handlebars-vscode).

**Comment by samselikoff** [comment]
@jgwhite I see, thanks. Do you know if it's possible to use an extension like that but also use `prettier#master` so I can try out these updates?

I did install master into my project and with the override above, `yarn prettier welcome.hbs` worked (it used the `glimmer` parser for my `hbs` file).

**Comment by jgwhite** [comment]
@samselikoff 🤔 I’m not sure. It’s a shame the extensions are architected that way.

**Comment by samselikoff** [comment]
Not exactly sure what I changed but https://github.com/mfeckie/handlebars-formatter seems to be using some of the new whitespace improvements.

Before:
![image](https://user-images.githubusercontent.com/2922250/67640066-10c7ed00-f8ce-11e9-8e37-4b5f122db297.png)

After:
![image](https://user-images.githubusercontent.com/2922250/67640073-17eefb00-f8ce-11e9-8816-31bc1ad97549.png)

Looks amazing, nice work everyone!

I'm also using the "Prettier - Code formatter" extension which was giving me trouble, so I had to add this to my `settings.json` to tell Code to use this extension for handlebars files instead:

```json
"[handlebars]": {
  "editor.defaultFormatter": "mfeckies.handlebars-formatter"
}
```

**Comment by rwjblue** [comment]
Not sure exactly what will need to be done, but wanted to point folks to some changes coming to `@glimmer/syntax` (which is what `prettier` uses, right?):

* https://github.com/glimmerjs/glimmer-vm/pull/983
* https://github.com/wycats/handlebars.js/pull/1584

This will significantly aid codemod utilities (`ember-template-recast` / `prettier`) by ensuring that the resulting AST includes things that had whitespace control.

**Comment by velrest** [comment]
Hey, 
just wanted to ask if this is intended behavior:
Before:
```hbs
<CardComponent
  @name="test"
  @size="l"
  @onclick={{true}}
  @color="default"
  as |card|
>
  <card.header>
    content
  </card.header>
</CardComponent>
```
After:
```hbs
<CardComponent
  @name="test"
  @size="l"
  @onclick={{true}}
  @color="default" as |card|
>
  <card.header>
    content
  </card.header>
</CardComponent>
```
The `as |x|` is always moved to the same line as the last component param. IMO this is not as readable as a new line but if this is intended im okay with it.

Also all new lines added for readability are stripped out. Is this also intended?

**Comment by jgwhite** [comment]
@velrest could I confirm which version of Prettier you’re using?

@chadian you’re probably the most familiar with intended Prettier behavior at this point. Could you check the above?

**Comment by jgwhite** [comment]
We’re still looking for an adventurous contributor to take on `{{! prettier-ignore }}` support.

**Comment by velrest** [comment]
@jgwhite im using prettier version `1.18.0`.

**Comment by velrest** [comment]
I just ran prettier with glimmer through a fairly large codebase and found some weird cases:

__Only closing tags on new line__
Before
```hbs
<UkButton
  @disabled={{or (if protectedPage (not warningRead)) protectedSystemPage}}
>
```
After:
```hbs
<UkButton
  @disabled={{or (if protectedPage (not warningRead)) protectedSystemPage
  }}
>
```
__Inserts newline into HTML attribute value if hbs variable is used__
Before:
```hbs
<span
  class="delete pointer show-delete-block-{{elementId}} uk-invisible uk-flex uk-flex-middle uk-flex-center"
>
```
After:
```hbs
<span
  class="delete pointer show-delete-block-
    {{elementId}}
    uk-invisible uk-flex uk-flex-middle uk-flex-center"
>
```
Without the `{{elementId}}` the line is not split.

Another example:
Before:
```hbs
<button
  class="uk-button uk-button-{{if @selected "primary" "default"}} {{unless @icon "round-button"}}"
>
```
After:
```hbs
<button
  class="uk-button uk-button-
    {{if @selected "primary" "default"}}

    {{unless @icon "round-button"}}"
>
```

__Newline inserted for every hbs statement__
Before:
```hbs
<span>
  ({{inc currentCardIndex}}/{{cards.length}})
</span>

```
After:
```hbs
<span>
  (
  {{inc currentCardIndex}}
  /
  {{cards.length}}
  )
</span>
```

It seems like the second and third issue are caused by enforcing that each hbs statement is on a new line.

**Comment by jgwhite** [comment]
@velrest this is great stuff to build test cases around!

Could you try Prettier 1.19.1 and see if any of this improves?

**Comment by velrest** [comment]
@jgwhite So the "Only closing tags on new line" and the "Newline inserted for every hbs statement" cases are resolved with prettier@1.19.1.

The "Inserts newline into HTML attribute value if hbs variable is used" is still a issue.

__All examples in the same order as before__
Before:
```hbs
<UkButton @disabled={{or (if protectedPage (not warningRead)) protectedSystemPage}} />
<span class="delete pointer show-delete-block-{{elementId}} uk-invisible uk-flex uk-flex-middle uk-flex-center" ></span>
<button class="uk-button uk-button-{{if @selected "primary" "default"}} {{unless @icon "round-button"}}" ></button>
<span>({{inc currentCardIndex}}/{{cards.length}})</span>
```
After:
```hbs
<UkButton
  @disabled={{or (if protectedPage (not warningRead)) protectedSystemPage}}
/>
<span
  class="delete pointer show-delete-block-
    {{elementId}}
     uk-invisible uk-flex uk-flex-middle uk-flex-center"
></span>
<button
  class="uk-button uk-button-
    {{if @selected "primary" "default"}}

    {{unless @icon "round-button"}}"
></button>
<span>
  ({{inc currentCardIndex}}/{{cards.length}})
</span>
```

**Comment by jgwhite** [comment]
@velrest awesome. Thanks so much for trying it out 👍

**Comment by dcyriller** [comment]
@velrest https://github.com/prettier/prettier/pull/7051 should fix issue 2 and 3.

**Comment by jgwhite** [comment]
Awesome, thanks @dcyriller @velrest

**Comment by amk221** [comment]
I have a large app I'd like to try this on, when I follow the instructions
`npm run prettier '**/*.hbs' --parser glimmer` it says `No parser could be inferred for file`
I presume I need to use this fork? How do I do that specifically?

**Comment by jgwhite** [comment]
@amk221 you shouldn’t need to use the fork (it doesn’t actually have any changes in it)

Could you try the config described here:
https://carakan.net/blog/2019/09/enable-prettier-in-ember-templates/

**Comment by amk221** [comment]
Thanks, that worked.

All ok apart from these two issues:

<img width="903" alt="Screenshot 2019-12-06 at 12 26 28" src="https://user-images.githubusercontent.com/382061/70323210-25c86200-1824-11ea-9f8e-ad6726437fa0.png">
<img width="858" alt="Screenshot 2019-12-06 at 12 29 12" src="https://user-images.githubusercontent.com/382061/70323212-26f98f00-1824-11ea-8c44-b85f9dbae16e.png">

**Comment by jgwhite** [comment]
@amk221 awesome, those are great corner cases to address.

**Comment by amk221** [comment]
np.

Only other things is:

```handlebars
<div>{{yield}}</div>
```

and

```handlebars
<div>
  {{~yield~}}
</div>
```

Always gets converted to:

```handlebars
<div>
  {{yield}}
</div>
```

So as gavin says, I will need:

```handlebars
<div>{{! prettier-ignore }}{{yield}}</div>
```

**Comment by amk221** [comment]
Oh and one more thing, `<br>` gets converted to `<br />`, which conflicts with `self-closing-void-elements` ember template lint rule. https://github.com/prettier/prettier/issues/5246.

So I'm not sure which to settle on. Go with Ember's recommended linting guidelines of `<br>`, or go with what Prettier does.

^ Would appreciate some advice on that one

**Comment by amk221** [comment]
Sorry for spam:

```handlebars
\{{escaped}}
```

Broke by becoming:

```handlebars
{{escaped}}
```

**Comment by Alonski** [comment]
@amk221 Regarding the Ember Template Lint rules that are conflicting with Prettier. The plan is to stop recommending these rules and to recommend using Prettier once it is ready.

So you can safely turn off those rules such as `self-closing-void-elements` and let Prettier do its job :)

**Comment by amk221** [comment]
Great, thank you!

**Comment by amk221** [comment]
Bit of a funky one here too:

<img width="942" alt="Screenshot 2019-12-06 at 19 35 23" src="https://user-images.githubusercontent.com/382061/70350938-b2444600-185f-11ea-9b69-82dc0587de59.png">

<img width="544" alt="Screenshot 2019-12-06 at 19 35 36" src="https://user-images.githubusercontent.com/382061/70350937-b1abaf80-185f-11ea-8f0a-03faa6ac809f.png">

(it breaks the class name `message--{{@type}}`)

**Comment by chrisrng** [comment]
Testing this out and it seems like even with Prettier 1.19.1 im seeing the yielded component formatting weirdly

![IMG_4922](https://user-images.githubusercontent.com/5466634/70377166-2c67df80-18df-11ea-8516-9150d954f437.JPG)

I would expect the `as |components|` to remain in a new line

**Comment by chrisrng** [comment]
Found that we strip the array brackets causing an issue.

Before
```hbs
{{some-component myprop=(some-helper someArray.[0])}}
```

After
```hbs
{{some-component myprop=(some-helper someArray.0)}}
```

Causing an error

**Comment by chrisrng** [comment]
We still seem to be stripping tildes from:

```hbs
{{~var~}}
```

to:

```hbs
{{var}}
```

**Comment by Alonski** [comment]
@chrisrng From what I remember of past discussions some Array.[0] shouldn't work and isn't valid hbs.
Which isn't great because it does work and people depend on it 😛
I changed all instances of usage like this up use an object-at helper or similar

**Comment by chrisrng** [comment]
@Alonski Any idea if this is documented anywhere? Actually shouldn't we just use the native get helper?

From:
```hbs
{{data.[0].someprop}}
```

To:
```hbs
{{get data "0.someprop"}}
```

**Comment by chrisrng** [comment]
From discord I guess the closest thing to this being documented is this issue: https://github.com/ember-template-lint/ember-template-lint/issues/787

**Comment by amk221** [comment]
```hbs
<a href="#">foo</a>
```
Results in:
```hbs
<a href="#">
  foo
</a>
```
which renders
<img src="https://user-images.githubusercontent.com/382061/71981838-1b322b80-321b-11ea-8118-b35ca4fd7048.png" width="49">

**Comment by chadian** [comment]
I took a try at prettier-ignore. Give the [pull request](https://github.com/prettier/prettier/pull/7275) a look or the [playground](https://deploy-preview-7275--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEAdKACDAeAJgSwDcsSwAbAQwGcqBeVEBGAJwE8GSMA+dLYYAIQYADszgwY+OMwC0+AOZQIYgL4reOABYBGLvyySYZOJzXYA9Dp6YcBYlnLU6DAEYRc7ENc4+S+rG4eviRq6BZ2XDjmESAANCAQwpLQVMigFMzMEADuAAoZCKkoFIQQ+LhxIC7MFGAA1uIAysK1+FDyyCwArnDxmjAAtmQA6pr48FQtYHCNheNE46zI4DSVbVTSMLk18gMUyABmFGQb8QBWVAAeAEI19U0UA3AAMm1wh8enIBeXjW3yxgAil0IPAPideiAWswNsxlgD8AMnnD4qI2jBhuUYJpkAAOAAMqKyG2GNWEy1EcFhhHe8QAjiD4NtEkUQNQZFA4HBcNzKmIGfgxNsKLt9kgjhD4hsBvhOswelL-kDGe9xZ9ITAKC5MbhscgAEzxFgUfBkf4AYQgSLFICUnMqXQ2ABUtUUJV9CD0AJJQHmwRpgZj4JIAQV9jRgrGM4I2aiAA) a try. I'm sure there's some edge cases I still need to account for.

**Comment by amk221** [comment]
<img width="425" alt="Screenshot 2020-01-15 at 09 20 05" src="https://user-images.githubusercontent.com/382061/72421431-621ea480-3778-11ea-9463-2353f7b1e88f.png">

...Inserted a line break into the URL.

When clicked however this still worked. Nonetheless, I don't think it should happen.

<img width="287" alt="Screenshot 2020-01-15 at 10 28 26" src="https://user-images.githubusercontent.com/382061/72426530-cdb93f80-3781-11ea-9077-9af53eca6e6b.png">

**Comment by dcyriller** [comment]
@amk221 a fix for that last issue was merged a few days ago https://github.com/prettier/prettier/pull/7051

**Comment by amk221** [comment]
Ah ok thank you. I should have checked.

**Comment by dcyriller** [comment]
No worries, it's not released yet. But best option might be to make tests on master

**Comment by gossi** [comment]
Repost from prettier/prettier#5340

Also ping @mfeckie, who can say more about the options that are being used.

----

I realized today, this is stripping the tilde character: `~`

**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACdqA8AbAlqgEwEMYiBaeAZxjICc4jLoA+AHSgw2GAGJcAzVEwC2cAHQwAFrigBzAL7z2nTvwi1U3AH4jxueMK2LlK7nGyU4xjivSZKAByIcw2RpQC8rEHGEOYAJ7ebDa26GQRZCa2mAD0js4httyxAtaccXjMIAA0IBD+uNCUyKBEtLQQAO4ACuUIJShEAG4QuAS5IABGtERgANZwMADKTmAyssgwtACucHmSMMLYAOrSVGNwww36uM36AcjglCV5Mpa0MDW9ssJEyPxEFvMgAFaUAB4AQr0DQ8NEUQAGRkcAeT0seXeH2GE2wcAAijMIPBwc88k5aBcjrI8MJRLROg5aDIYCt2lJkAAOAAMGMqlhWvQcR2JcAuzTBeQAjsj4NcCo0QIwyFA4HACBLOvRebh6NciLd7khHuiQJZhLgprMXpQ4Yi+WCVRCXqQuuSCJSkAAmPLTIi4PByADCEHxypAUGgXJAM0sABUiF1GqrISBmnMAJJQSWwYZgEn+ACCMeGgXhaMsiiAA)
```sh
--parser glimmer
```

**Input:**
```hbs
    <li data-test-reason>
      {{#if some.thing}}
        for {{~some.item~}}
      {{else}}
        <span class="empty">
          ----
        </span>
      {{/if}}
    </li>
```

**Output:**
```hbs
<li data-test-reason>
  {{#if some.thing}}
    for{{some.item}}
  {{else}}
    <span class="empty">
      ----
    </span>
  {{/if}}
</li>
```

Original Report: mfeckie/handlebars-formatter#18

**Comment by amk221** [comment]
```hbs
{{@foo}}.
```
is becoming:
```hbs
{{@foo
}}.
```

**Comment by amk221** [comment]
Should I be raising separate issues for these?

**Comment by jgwhite** [comment]
> Should I be raising separate issues for these?

@amk221 I think this thread is probably the best place for now.

**Comment by amk221** [comment]
Does anybody know what the 'solution' this one is: https://github.com/jgwhite/prettier/issues/1#issuecomment-572051505
...whilst still satisfying ember-template-lint?

**Comment by chadian** [comment]
We can update:
- [x]  Add support for `{{! prettier-ignore }}` (https://github.com/prettier/prettier/pull/7275)
It's been merged!

**Comment by dcyriller** [comment]
@amk221 you’re probably looking for this https://github.com/prettier/prettier/pull/7306#issuecomment-574238437

**Comment by amk221** [comment]
@dcyriller thanks!

**Comment by FabHof** [comment]
I just opened prettier#6464

**Comment by chrisrng** [comment]
> Found that we strip the array brackets causing an issue.
> 
> Before
> 
> ```
> {{some-component myprop=(some-helper someArray.[0])}}
> ```
> 
> After
> 
> ```
> {{some-component myprop=(some-helper someArray.0)}}
> ```
> 
> Causing an error

Mahir made a codemod to deal with this problem https://github.com/mahirshah/ember-array-index-template-codemod

**Comment by mahirshah** [comment]
I just opened [prettier#7575](https://github.com/prettier/prettier/pull/7575) to preserve whitespace stripping in output ("~")

**Comment by edflips** [comment]
I've been following development on this, and it's shaping up really nicely!

There is an issue when formatting repeatedly creates extra blank lines when an inline `if` is inside quotes.

```
local-class="some-class {{if (contains date.date this.someDates) "active"}}"
```

becomes

```
local-class="some-class
	{{if (contains date.date this.someDates) "active"}}"
```

and then

```
local-class="some-class

	{{if (contains date.date this.someDates) "active"}}"
```

etc.

Interestingly, it stops adding lines after a few formats.

**Comment by FabHof** [comment]
> There is an issue when formatting repeatedly creates extra blank lines when an inline `if` is inside quotes.

I just updated my PR prettier#6464 to fix this as well.

**Comment by FabHof** [comment]
https://github.com/prettier/prettier/pull/6464#discussion_r387681207 
what to do about that?🤔

**Comment by FabHof** [comment]
Current master turns

```hbs
<div title="Cool
{{stuff}}
and more"
>
  Hello
</div>
```

into 

```hbs
<div
  title="Cool
  {{stuff}}
   and more"
>
  Hello
</div>
```

Which renders a different tooltip.

**Comment by Alonski** [comment]
Just wanted to jump in with this crazy div:
```hbs
<div
  data-test-item={{@item.id}}
  class="list-item w-full group flex justify-between items-center text-base h-12
    {{if @isGroup "cursor-auto" "cursor-pointer hover:text-dark"}}

    {{if (and this.isSelected (not @isGroup)) "item-selected" "text-light"}}

    {{if this.isSelected "font-semibold text-dark" "text-light"}}"
  role="button"
  {{action "selectItem"}}
>
```
The empty newlines are added by prettier :)

Also it is impossible `{{! prettier-ignore }}` this :) The line just jumps around.
Before:
```hbs
<div
{{! prettier-ignore }}
  class="list-item w-full group flex justify-between items-center text-base h-12


    {{if @isGroup "cursor-auto" "cursor-pointer hover:text-dark"}}



    {{if (and this.isSelected (not @isGroup)) "item-selected" "text-light"}}



    {{if this.isSelected "font-semibold text-dark" "text-light"}}"
  role="button"
  {{action "selectItem"}}
>
```
After:
```hbs
<div
  class="list-item w-full group flex justify-between items-center text-base h-12


    {{if @isGroup "cursor-auto" "cursor-pointer hover:text-dark"}}



    {{if (and this.isSelected (not @isGroup)) "item-selected" "text-light"}}



    {{if this.isSelected "font-semibold text-dark" "text-light"}}"
  role="button"
  {{action "selectItem"}}
  {{! prettier-ignore }}
>
```

**Comment by Alonski** [comment]
Also just got this :)
Before:
```hbs
<Textarea />
{{#if}}{{/if}}
```
After:
```hbs
<Textarea />

{{#if}}{{/if}}
```
Again?
```hbs
<Textarea />
{{#if}}{{/if}}
```

So it seems this is non deterministic :)

**Comment by dcyriller** [comment]
Current master is now a bit out to date. They are currently working on the `next` branch to cut a v2 release. And bug fixes have been merged on that branch.

So @Alonski your first issue would be fixed there (even though running the snippet you posted highlights another bug 😅).

@FabHof the issue you posted is still there and a good candidate to fix :)

**Comment by FabHof** [comment]
@dcyriller but how to fix this? The only options I can think about is to

 - not do anything at all with attributes or
 - whitelist attributes that should be formatted (which are those? only class, or is there more? style maybe?)

**Comment by Alonski** [comment]
This is still an issue with Prettier v2.0.1 :)
https://github.com/jgwhite/prettier/issues/1#issuecomment-595098884
Before:
```hbs
<Textarea />
{{#if true}}
  Test
{{/if}}
```
After:
```hbs
<Textarea />

{{#if true}}
  Test
{{/if}}
```
Again?
```hbs
<Textarea />
{{#if true}}
  Test
{{/if}}
```

**Comment by alexlafroscia** [comment]
I'm seeing a similar case as @Alonski -- just one file (out of about 500 in our repo) is also non-deterministically entering and removing a newline after a angle-bracket component with each Prettier run.

For context, I'm trying with Prettier `2.0.2` right now after having tried with an older Prettier version and seeing even more issues.

**Comment by amk221** [comment]
What does everybody do about:

```hbs
<p>
  final paragraph
  <strong>the end</strong>.
</p>
```

becoming

```hbs
<p>
  final paragraph
  <strong>
    the end
  </strong>
  .
</p>
```
```
final paragraph the end .
                       ^ space
```

Just prettier ignore each time?

**Comment by samselikoff** [comment]
Not sure if this is a prettier thing or a code thing but when I comment out a selection like

```hbs
<div class="p-4 text-sm font-medium text-gray-500">
  Lorem ipsum dolor sit amet, consectetur adipisicing elit. Mollitia odit quod ab, asperiores, praesentium veniam.
</div>
```

vscode turns it into

```hbs
{{!-- <div class="p-4 text-sm font-medium text-gray-500">
  Lorem ipsum dolor sit amet, consectetur adipisicing elit. Mollitia odit quod ab, asperiores, praesentium veniam.
</div> --}}
```

and then when I save prettier turns it into

```hbs
{{! <div class="p-4 text-sm font-medium text-gray-500">
  Lorem ipsum dolor sit amet, consectetur adipisicing elit. Mollitia odit quod ab, asperiores, praesentium veniam.
</div> }}
```

Then when I press +/ to uncomment it, code gets confused and wraps it in another `{{!--` style comment

```hbs
{{!-- {{! <div class="p-4 text-sm font-medium text-gray-500">
  Lorem ipsum dolor sit amet, consectetur adipisicing elit. Mollitia odit quod ab, asperiores, praesentium veniam.
</div> }} --}}
```

**Comment by Alonski** [comment]
@samselikoff Also what is your current workaround :)

**Comment by samselikoff** [comment]
Yes thx for reminding me!

For now I Open Keyboard Shortcuts (JSON) and added this to the bottom:

```js
  {
    "key": "cmd+/",
    "command": "editor.emmet.action.toggleComment",
    "when": "editorLangId == handlebars"
  }
```

This relies on Emmet being installed. It will use HTML comments `<!--` in handlebars templates, which don't get formatted by prettier + don't confuse code.

**Comment by gvocale** [comment]
> What does everybody do about:
> 
> ```
> <p>
>   final paragraph
>   <strong>the end</strong>.
> </p>
> ```
> 
> becoming
> 
> ```
> <p>
>   final paragraph
>   <strong>
>     the end
>   </strong>
>   .
> </p>
> ```
> 
> ```
> final paragraph the end .
>                        ^ space
> ```
> 
> Just prettier ignore each time?

I manually have to go in and fix it, when I catch it 😢

**Comment by amk221** [comment]
@gvocale yep, same.

Problem with that is, each time you run `prettier —write`. It breaks them again.

And I don’t fancy writing tests for the expected placement of full stops! :)

**Comment by nightire** [comment]
I found a bug in Prettier v2.0.x:

```handlebars
{{! correctly formatted on top-level }}
{{something
  foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
}}

{{! correctly formatted inside a handlebars block expression }}
{{#let
  (or
    this.reallyLongPropertyToEnforceBreakIntoMultipleLines @reallyLongPropertyToEnforceBreakIntoMultipleLines
  ) as |reallyLongPropertyToEnforceBreakIntoMultipleLines|
}}
  {{something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
{{/let}}

{{! incorrectly formatted inside a regular HTML tags }}
<div>
  {{
    something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
</div>

{{! incorrectly formatted inside a Angle Bracket Component }}
<SomeComponent>
  {{
    something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
</SomeComponent>
```

As above snippet suggests, long enough inline expression will be formatted incorrectly if it is nested in any HTML tags or Angle Bracket Component invocations.

Using Prettier v1.19.x will result correctly:

```handlebars
{{something
  foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
}}

{{#let
  (or
    this.reallyLongPropertyToEnforceBreakIntoMultipleLines @reallyLongPropertyToEnforceBreakIntoMultipleLines
  ) as |reallyLongPropertyToEnforceBreakIntoMultipleLines|
}}
  {{something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
{{/let}}

<div>
  {{something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
</div>

<SomeComponent>
  {{something
    foo=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
    bar=this.reallyLongPropertyToEnforceBreakIntoMultipleLines
  }}
</SomeComponent>
```

**Comment by kangax** [comment]
Hey folks. Love this effort and would love to help with what I can. 

I've only recently started working on an Ember app and was sad to find out there's no prettier support for handlebars. Am I understanding that this will require Prettier 2.0 since the fixes are not being back-ported to 1.x?

**Comment by mydea** [comment]
Just tried this on our app, and it worked quite well. The only thing that is pretty much a dealbreaker for us is that it removes _all_ empty lines, making the code unnecessarily dense. Is that on purpose, or is there a way to deactivate this, and make it behave like the JS parser (which collapses empty lines to max. 1 line)? I looked through the issue and comments a bit but it's a bit hard to find anything specific :)

Another piece of feedback:

I think it would be nice to follow the JS handling of objects here, where if you manually put an object on multiple lines, it remains like that.

Currently, if I have this template:

```hbs
{{yield
  (hash
    name='X'
    other='Y'
  )
}}
```

It will turn this into:

```hbs
{{yield (hash name='X' other'Y')}}
```

Which is IMHO not ideal for Git versioning etc. (e.g. if I add a property it is much harder to see what changed etc.). The JS parser will respect such manual muli-line objects, and I think it makes sense to behave the same way here.

And finally, what is the reasoning behind putting the `as |...|` part on the same line as the last argument? E.g. I think this:

```hbs
<MyComponent
  @prop={{true}}
  @prop2={{true}}
  as |thing|
>
```

Would make more sense than what I currently get:

```hbs
<MyComponent
  @prop={{true}}
  @prop2={{true}} as |thing|
>
```
Anyway, thanks for the effort, I really appreciate it, and mostly it is working really well for me :+1:

**Comment by nightire** [comment]
@mydea Actually Prettier will preserve at most one intentional blank line. I guess you formatted your code by using this extension: https://github.com/Alonski/prettier-for-handlebars-vscode in vscode because it's which been recommended nowadays.

However, IMO it's not as good as another one in practice: https://github.com/mfeckie/handlebars-formatter. For example, the latter will preserve the intentional blank line but the former will not.

They both use the Prettier glimmer parser. Since there's no user config that we can change, the blank line preservation must be done by the extension itself, not the Prettier. In fact, I've tested this behavior by executing the prettier in CLI, it turns out will not remove the intentional blank line at all.

I'm not sure why people always recommended the former, maybe because of its activity, ~but it's will not be recommended by me currently.~

---

EDIT: To understand the reason behind the difference between these two extensions, see here: https://github.com/Alonski/prettier-for-handlebars-vscode/issues/9#issuecomment-609669882

**Comment by mydea** [comment]
I formatted my files via: `prettier --write "**/*.hbs" --parser glimmer`, which also removes the lines.

It _does_ work if I use the handlebars-formatter extension, though!

(I am using Prettier 2.0.2, by the way)

**Comment by Alonski** [comment]
@mydea @nightire Please note both formatting extensions for VSCode use a set Prettier version. If you format via the Prettier CLI it will format differently :)

**Comment by mydea** [comment]
> @mydea @nightire Please note both formatting extensions for VSCode use a set Prettier version. If you format via the Prettier CLI it will format differently :)

Hmm, good to know. But what leads to the different behavior in blank line handling then? With the CLI (and Prettier 2.0.2), as well as with prettier-for-handlebars-vscode, all blank lines are completely removed, while handlebars-formatter keeps one blank line for me.

**Comment by EvgenyOrekhov** [comment]
Are there plans to add support for `hbs` template tag? Like in Ember tests, for example:

```js
await render(
  hbs`<MyComponent @foo={{this.foo}} />`
);
```

**Comment by kangax** [comment]
Just ran `prettier --write "**/*.hbs" --parser glimmer` over our codebase (using Prettier 2.0.5) and here are some things I'm seeing:

1. Comments changing to single-line ones (from `{{!--` to `{{!`). Can we control this behavior?
2. End of line is removed if (I assume) the file ends with a directive (e.g. `</unless>`)
3. 1 attribute with a long value is moved to the new line but shouldn't (at least that's the JSX behavior and I agree it looks better):
```hbs
<div
  class="wvmp-v2-profile-viewer-ghost-card-mobile module padding-20 display-flex flex-column align-items-center"
>
```
4. <del>Component name is sometimes incorrectly moved to the new line</del> (NOW FIXED):
```hbs
{{
  abi$abi-form
  acceptInvite=model.acceptInvite
  sendInvite=model.sendInvite
  profile=model.profile
  ...
}}
```
5. Something going on with self-closing tags. `<br>` becomes `<br />`, is this expected?
6. Line width should not take control characters into consideration (or, at least, not move them to the new line):
```hbs
{{image-view-model$image-view-model images=actor.image isPresenceEnabled=true entitySize=3
}}
```
7. <del>Ignoring certain helpers by default. We have `t` helper for localization, and I imagine, some people would not be OK with something like this</del> (NOW FIXED):
```hbs
{{   
    t
    "i18n_..."
    date=(ember-cli-...$format-date
        publishedAt format="..." useTimeZone=true
    )
}}
```
`t` being on a new line (from `{{`) also breaks highlighting in my VSCode.

Hope this helps. If there is consensus on any of these being incorrect, I can look into fixing them.

**Comment by mydea** [comment]
I agree with everything you said @kangax!

**Comment by GCheung55** [comment]
I’m interested in helping get Glimmer support across the finish line. How can I help?

**Comment by jgwhite** [comment]
> I’m interested in helping get Glimmer support across the finish line. How can I help?

@GCheung55 thanks for your interest.

The two primary ways to help out:

1. Run prettier on your own codebase(s) and report anything that makes you frown (see description for details)
2. We’ve fallen a little behind on keeping the checklist up-to-date, but take a look at reports such as https://github.com/jgwhite/prettier/issues/1#issuecomment-630412651 and see if you can submit a PR to Prettier that fixes it.

If you want a hand getting up-and-running with Prettier, let us know here and we’ll see if someone has time to pair.

**Comment by bartocc** [comment]
With prettier 2.0.5, we ran into an issue when trying to format the following `index.hbs` file:

```hbs
Start a <span>télé-</span>consultation
```

- output with `prettier --parser html index.hbs`:

```hbs
Start a <span>télé-</span>consultation
{{!-- template untouched --}}
```

- output with `prettier --parser glimmer index.hbs`:

```hbs
Start a
<span>
  télé-
</span>
consultation
{{!-- a space will be rendered between `télé-` and `consultation`. --}}
```

I believe this is a bug because I don't see any reason why the `glimmer` parser should generate a different output from the `html` parse.

For the time being, our only workaround is to wrap the template's content in a div tag and disable prettier for this div

```hbs
{{! prettier-ignore }}
<div>
  Start a <span>télé-</span>consultation
</div>
```

cc: @pbernery @monsieurdart @pauljeannot @noctambul

**Comment by stfnio** [comment]
> t being on a new line (from {{) also breaks highlighting in my VSCode.

I think that's the only thing that stops me of using "format on save".

**Comment by dcyriller** [comment]
> 2. We’ve fallen a little behind on keeping the checklist up-to-date, but take a look at reports such as #1 (comment) and see if you can submit a PR to Prettier that fixes it.

To follow-up on what @jgwhite wrote: an intermediate step can be to open an issue in Prettier repo. It is very easy to do thanks to https://prettier.io/playground where you can copy-paste your code and click the `Report issue` button. The support for Glimmer parser was recently added to the playground.

**Comment by dcyriller** [comment]
@kangax 
1/ For now, any comment that includes a mustache is formatted with `{{!-- --}}` and any comment _not_ including a mustache with `{{! }}`. Thus, the formatted code remains correct regarding ([handlebars documentation for comments](https://handlebarsjs.com/guide/#template-comments)). But your question was: why this formatting instead of respecting the user choice for one syntax or the other? I can't tell for sure because I don't have the whole history, but I think that the problem is that these two syntaxes have an exact same AST representation. So when we print them, we don't know if the user used the former or the latter syntax. 

That said, I've not yet figured out why we have two different comment syntaxes in handlebars 😄. They have the same AST. The only difference I see between the two syntaxes is that the former supports including mustache. So, should we converge on formatting every comment to `{{!-- --}}`? _EDIT: the topic is already discussed [in this RFC](https://github.com/emberjs/rfcs/issues/504)_

2/ At the moment, Prettier will always remove the final newline, see [this snapshot](https://github.com/prettier/prettier/blob/master/tests/handlebars/last-line/__snapshots__/jsfmt.spec.js.snap) (the snapshots are formatted to be human readable).

On that topic, Prettier complies with the choice made in [ember-cli blueprints](https://github.com/ember-cli/ember-cli/blob/v3.19.0-beta.2/blueprints/app/files/.editorconfig#L16). Why removing the final newline? Because it inserts a trailing `TextNode` which is useless and sometimes can break the UI.

We could change that and try to respect the user's choice made through editorconfig's `insert_final_newline` option. To do so, we'd have to add support for `insert_final_newline` option in https://github.com/josephfrazier/editorconfig-to-prettier. I'm not sure if they'd be willing to add it though (usually reluctant to adding more options). If they're not ready to accept it, we'd be back to the choice made in ember-cli and if we want to change that choice, we could open a RFC.

3/ and 6/: I agree the formatting of these snippets doesn't look good. Should we open issues?

4/ and 7/ is the same issue. It was not formatted this way in previous versions, it is a regression (@nightire shows it a few comments above). We could open an issue to track the progress made on it.

5/ I think this is expected for html _for now_ (so for handlebars as well). There is a discussion opened on this topic in https://github.com/prettier/prettier/issues/, 5246.

@EvgenyOrekhov 
> Are there plans to add support for hbs template tag? 

In the javascript printer there is a `embed` hook. It could be used to format some handlebars code (embed in javascript code) with Glimmer printer. Should we open an issue to track that?

@mydea 

> Which is IMHO not ideal for Git versioning etc. 

I agree.

Also, I opened https://github.com/prettier/prettier/issues/8502

@Alonski 
I reported this issue here: https://github.com/prettier/prettier/issues/8504 :)

**Comment by EvgenyOrekhov** [comment]
@dcyriller 

> Are there plans to add support for `hbs` template tag?

> Should we open an issue to track that?

Yes, please!

**Comment by stfnio** [comment]
Just came up with this weird [result](https://upload.visible.vc/4gujbj6d) of formatting a file. Reproducible by having a string which starts with `\n` and continues on the next line. Could you add this to the list of knows issues?

**Comment by amk221** [comment]
What's recommended for a situation like this:

```hbs
{{! in }}
<FormField>Field Name<Required /></FormField>
```

```hbs
{{! out }}
<FormField>
  Field Name
  <Required />
</FormField>
```

```hbs
{{! desired }}
Field Name*
```

```hbs
{{! actual }}
Field Name *
```

**Comment by amk221** [comment]
```handlebars
{{! input }}
\{{my-component}}
```

```handlebars
{{! output }}
{{my-component}}
```

..this causes the my-component to be rendered, when it shouldn't.

**Comment by salomvary** [comment]
In:
```handlebars
<!DOCTYPE html>
<html>
<body>
    {{{body}}}
</body>
</html>
```

Out:
```handlebars
{{{body}}}
```

Works as expected without `<!DOCTYPE html>`. See in [playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAhAEQPIGEAqAmgAoCiABABYwC2ANgHwA6Uq19zrARhACYCencsPLAxPAQF9pLVAHoJg2XPaMQAGhAQADjACW0AM7JQAQwBO5iAHdiFhMZSm610-2Oau502ADWcGABlbR89KABzZBhzAFc4TVUAdUo9eEMQsDhAh1S9ADdU-mRwQw8QMMM4cxhib3CaU2QAM2dKzQArQwAPACFvPwDA0xo4ABkwuGbW+JBOrsCw8Lo4AEUYiHgpujaQEPNK82KlvRoRw81tczCYRL1eGEpkAA4ABgurSsTvbWLLuAO8pNNABHdbwWo6RwgUyGAC0UDgcF4SI0IHMcFBenRtVM9UaSBa2xmlRoeiisWJi2Waw2kwJ000MFMXFu90eSAATIzvHo6ItcBBTviQP8AKyomKVfDMxyEnZ5OIASSgyNggTAV10AEEVYEYPxlltKtIgA).

**Comment by kangax** [comment]
@salomvary I looked into the DOCTYPE issue and it seems that glimmer parser doesn't recognize it in the first place. It's not even in the AST so prettier can't process it. Looks like this is a known issue and isn't something prettier can fix until glimmer adds support for it — https://github.com/glimmerjs/glimmer-vm/issues/870

**Comment by rwjblue** [comment]
I had a PR for it, but ended up closing 🤔

**Comment by kangax** [comment]
@bartocc @amk221 the incorrect linebreak in your examples is a known issue (and is probably one of the major ones since it affects rendering) — https://github.com/prettier/prettier/issues/8527 

@dcyriller I was looking more into 3rd issue in my https://github.com/jgwhite/prettier/issues/1#issuecomment-630412651 and realised that prettier is inconsistent in how it handles this. JSX does not break if it's 1 attribute with long value while both HTML and Glimmer printers do. Now I'm not sure what it should be, we can consider this feature not a bug :)

@jgwhite could you update the checklist for this ticket? looks like all of the issues listed are now fixed. Should we instead just link to [bugs labeled with "lang:handlebars"](https://github.com/prettier/prettier/issues?q=is%3Aopen+label%3Atype%3Abug+label%3Alang%3Ahandlebars)?

**Comment by jgwhite** [comment]
> @jgwhite could you update the checklist for this ticket? looks like all of the issues listed are now fixed. Should we instead just link to bugs labeled with "lang:handlebars"?

@kangax that is a terrific idea, done 👍

**Comment by amk221** [comment]
There's an issue with `{{! prettier-ignore }}`

If you want a file with _no whitespace in it_:

`{{! prettier-ignore }}<div>{{@text}}</div>`

Results in:

```
{{! prettier-ignore }}
<div>
  {{@text}}
</div>
```

<hr />

What's the situation with thread now? Should we still be raising issues here, or do it on Prettier repo proper?

**Comment by nabeelz6** [comment]
>rwjblue commented on Jun 19, 2019
RE: the entity escaping, we really need to avoid the escaping all together (see prettier#4533 (comment) for more details)

>GavinJoyce commented on Jun 19, 2019
I'm working on fixing the HTML entities here: prettier#6234

So while HTML entity names (e.g. `&lt;`) are not being converted since this fix, It seems that prettier via glimmer parser is still converting HTML entity numbers (decimals) via glimmer parser which doesn't seem right. 

Before (see [playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBBATnABFCMsC2EAJgJYBmpcxWYAFgIboNjzoDOSWAZADYwBuHgHNBPBgQAOAgDpQM2ImUrVajZqzgcu3AMQA2AAxC9AJgCcpoSAA0ICJJilo7ZKCboIAdwAKTBK4oDLxeDACernYARhoA1nAwAMqSLKRQwsgw6ACucHZwBFHUxNQAMgzp2QzCcABiEOgEDDBO6cggDNkwELYgdDAEvADqdKTw7ClgcIkBY6QAbmNh7exZpKy9aexaMD7Mwk3I5MHbdgBW7AAeAEJxCYkScKVpcEcneSAXl4lpwrxwAEVsnhXkhjrxTiAUhwtO0-qQCARYXZJOg0jAhqRiDA6MhTIYUZ5tkNmJJ2qi4Nt0PNXnYAI7A+B7ByBDrsAC0UDg1GovUwDNImD21UOYPedm2BFImRyH3Yv3+QJBbwhHxgDCimOxuKQpjsWQYpF4vwAwhBEQx2pSAKy9bLbAAqGsC4Mh81yAEkoCVYIkwGjHKhvYkYGF-irtgBfSNAA)):
```
Are not modified characters: &lt; &gt; &amp;
Are modified characters: &#60; &#292; 
```

After:
```
Are not modified characters: &lt; &gt; &amp;
Are modified characters: < Ĥ
```

We have a fairly large codebase using HTML entity numbers and while converting some of the numbers won't break anything (except maybe some older browsers), it does break for things like `&#60;`. As a workaround, we could manually replace all instances with their equivalent names but it does feel like prettier shouldn't be messing with HTML entities period.

Should I add a bug for this in prettier repo?  Btw, current glimmer behavior is also inconsistent with JSX parser (see https://github.com/prettier/prettier/issues/678).

References:
- https://github.com/prettier/prettier/pull/4533
- https://github.com/prettier/prettier/pull/6234
- https://github.com/glimmerjs/glimmer-vm/pull/938

**Comment by rwjblue** [comment]
@nabeelz6 - Mind making a failing test case for glimmerjs/glimmer-vm for this? Off the top of my head, I don't know where those are being processed but a test case should help us get started.

**Comment by dcyriller** [comment]
I opened a PR to have a whitespace-sensitive mode for Prettier. It could be a solution to the whitespace issue around `span` and other inline elements. 

I ran the code against the ember-observer code base. Any feedbacks on how the templates are formatted is welcome in the PR: https://github.com/dcyriller/client/pull/2.

**Comment by thorn0** [comment]
So, Glimmer folks, now that we have `--html-whitespace-sensitivity`, does anything else still separate the Glimmer parser from being out of experimental?

**Comment by chrisrng** [comment]
Agree ^ anyone know the maintainers to see if we can cut a release?

**Comment by ghost** [comment]
@thorn0 @dcyriller a huge thank you for landing Handlebars support!
I just started to read the [official blog post goodies](https://prettier.io/blog/2021/05/09/2.3.0.html#move-handlebars-support-out-of-alpha-to-release-10290httpsgithubcomprettierprettierpull10290-by-dcyrillerhttpsgithubcomdcyriller--thorn0httpsgithubcomthorn0)

**Comment by jgwhite** [comment]
I’d say we can close this issue now 🎉
